android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+                throw new TimeoutException("onBind - unable to acquire startup lock within timeout");
+            }
+            // Proceed
+            Log.d(LOG_TAG, "onBind - lock acquired");
+            MessengerLooperThread msgThread = messengerThread.get();
+            if (msgThread != null) {
+                return msgThread.getMessengerBinder();
+            } else {
+                return null;
+            }
+        } catch (InterruptedException ignore) {
+            Log.d(LOG_TAG, "onBind - lock wait interrupted");
+            return null;
+        } catch (Throwable e) {
+            Log.e(LOG_TAG, "Exception " + Log.getStackTraceString(e));
+            return null;
+        } finally {
+            if (hasEnvironment()) {
+                new Handler().postDelayed(() -> {
+                    try {
+                        startInForeground();
+                    } catch (Throwable e) {
+                        Log.d(LOG_TAG, "Error while starting in foreground - " + e);
+                    }
+                }, 500);
+            }
+            if (startupLockAcquired) {
+                serviceStartupLock.release();
+                Log.d(LOG_TAG, "onBind - lock released");
+            }
+        }
+    }
+
+    /**
+     * Set and get old thread join time.
+     *
+     * @param joinTime new join time
+     * @return old join time
+     */
+    protected long getAndSetJoinTime(long joinTime) {
+        return threadJoinTime.getAndSet(joinTime);
+    }
+
+    /**
+     * Gets component's arguments.
+     *
+     * @return array of component's arguments.
+     */
+    protected String[] getArguments() {
+        return arguments;
+    }
+
+    private void startInForeground() {
+        Notification notification = getNotification();
+        startForeground(getNotificationId(), notification);
+    }
+
+    private void start() {
+        Log.d(LOG_TAG, "start");
+
+        Thread thread = new Thread(() -> {
+            exitCode = EXIT_CODE_FAILED;
+            PrintStream stdout = null;
+            PrintStream stderr = null;
+            try {
+                setRunningState();
+                Log.d(LOG_TAG, "Service thread started");
+                // Override stdout and stderr to send lines to Nucleus
+                stdout = System.out;
+                stderr = System.err;
+                System.setOut(new PrintStream(new GreengrassOutputStream(LIFECYCLE_MSG_STDOUT_LINES,
+                        LIFECYCLE_EXTRA_STDOUT_LINE)));
+                System.setErr(new PrintStream(new GreengrassOutputStream(LIFECYCLE_MSG_STDERR_LINES,
+                        LIFECYCLE_EXTRA_STDERR_LINE)));
+
+                onServiceStart();
+                if (componentWorkerThread != null) {
+                    componentWorkerThread.start();
+                    componentWorkerThread.join();
+                    exitCode = componentWorkerThread.getExitCode();
+                }
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)